[dynamo][moco] Save global torch state to restore on graph break

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101201


This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

anijain2305 committed:
[dynamo] Save global torch state to restore on graph break

[ghstack-poisoned]

voznesenskym commented on torch/_guards.py:
```diff
@@ -344,6 +344,52 @@ def restore_graphstate(self, state):
+    def __init__(self):
+        self.global_state: Dict[str, bool] = {}
+
+    def copy_graphstate(self):
+        return GlobalContextCheckpointState(dict(self.global_state))
+
+    def restore_graphstate(self, state):
```
These need to be added to the copy/restore flow on TX

anijain2305 replied:
Can you elaborate more? I am calling these during the copy/restore in the `output_graph`

voznesenskym replied:
Oh I missed that this is in output graph. That works for me! 

Consider moving it up to Tx, it fits there better, but no strong opinions.

voznesenskym commented:
Can you reference https://github.com/pytorch/pytorch/pull/100570 in your description? A lot of relevant conversation there.

anijain2305 commented:
@pytorchbot merge

bdhirsh commented on torch/_dynamo/output_graph.py:
```diff
@@ -354,6 +362,10 @@ def guards(self) -> Set[Guard]:
 
+    def save_global_state(self):
+        global_state = self.tracing_context.global_context.global_state
+        global_state["is_grad_enabled"] = torch.is_grad_enabled()
```
basic question: there's other global state, like e.g. current autocast context. Should this function be checkpointing that too?

anijain2305 replied:
Yes, thats what I was thinking too. Yes, we should. Thats why I created this central place to handle global states. We will add more cases here as they come.

anijain2305 committed:
Update on "[dynamo][moco] Save global torch state to restore on graph break"



This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -f "unrelated failures"

anijain2305 committed:
Update on "[dynamo][moco] Save global torch state to restore on graph break"



This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][moco] Save global torch state to restore on graph break"



This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][moco] Save global torch state to restore on graph break"



This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -f "expecting the only remaining job to pass, merging with -f to avoid rebasing with frequently changed file"
